---
title: "Exploring the results of omics data analysis inetractively"
author: "Alex Sanchez"
date: "December 2, 2016"
output: 
  html_document:
    toc: true
    toc_float: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

#Introduction

Omics data analysis usually yield its results as a variey of tables and graphs.

It often happens that we need to review or reanalyze some of these results making plots or computations on a  subset of selected features.

The goal of this document is to illustrate how to do some reanalyses using R code. Ideally some of these may -in a future work- be implemented in some type of interactive application that allows the researcher to explore the data him/herself. 

## A typical scenario and a wishlist

A typical omics data analysis is described in the documents available in this github repository:  
[https://github.com/alexsanchezpla/scripts/tree/master/Exemple_Analisis_BioC](https://github.com/alexsanchezpla/scripts/tree/master/Exemple_Analisis_BioC).

Usual input/output from this type of analysis may be:

- Input
    - A gene expression matrix or similar:   [https://raw.githubusercontent.com/alexsanchezpla/scripts/master/Exemple_Analisis_BioC/results/Datos.Normalizados.Filtrados.csv2](https://raw.githubusercontent.com/alexsanchezpla/scripts/master/Exemple_Analisis_BioC/results/Datos.Normalizados.Filtrados.csv2)
    - A `targets` file with the covariates for the analysis:  
[https://raw.githubusercontent.com/alexsanchezpla/scripts/master/Exemple_Analisis_BioC/data/targets.txt](https://raw.githubusercontent.com/alexsanchezpla/scripts/master/Exemple_Analisis_BioC/data/targets.txt)    
    - A `topTable` such as that produced by the `limma` Bioconductor package:    
    [https://raw.githubusercontent.com/alexsanchezpla/scripts/master/Exemple_Analisis_BioC/results/ExpressAndTop_AvsB.csv2](https://raw.githubusercontent.com/alexsanchezpla/scripts/master/Exemple_Analisis_BioC/results/ExpressAndTop_AvsB.csv2).  
    - Notice that the first two correspond to analysis inputs whilst the `topTable` is an analysis output, although it is an input for the "reanalysis".
    
- Output
    - Given a gene (or a set of genes) we may want, for each of them
        - Establish if the expression of the gene is related to a given categorical variable
            - Boxplot/dotplot
            - ANOVA
        - Establish if the expression of the gene is related to a given continuous variable
            - Scatterplot
            - Correlations
    - Given a subset of genes (and samples) from the expression matrix we may want: 
        - To investigate the correlation between these genes, either graphically or numerically.
        - To draw a heatmap on the selected rows/columns, perhaps to highlight some characteristic that the group has or simply to do this with a distinct colour.
        - To compute and plot a `Principal Components Analysis` on the data, perhaps to investigate the effect of removing one or more samples.
    - Given a _topTable_ (ordered results) we may want: 
        - To draw a volcano plot where we may wish to interactively highlight some of the points.
        - To do a biological significance analysis on the differentially expressed genes.
    - The results of these reanalyses may be sent to xls/html/pdf/png files or shown in the screen.
    -*Notice that these points are related. Very often the subset of genes will be determined by the values in the _topTable_*
            
The code in the following sections illustrates some examples on how one could do these analyses (obviously this can always be improved). Because one of the goals of this document is to serve as the basis for an interactive document the examples distinguish the parameters from the execution of each call.


### Setting the environment

```{r setDirs}
workingDir <- getwd()
dataDir <- file.path(workingDir, "dades")
codeDir <- file.path(workingDir, "Rcode")
resultsDir <- file.path(workingDir, "results")
setwd(workingDir)
```

```{r loadpackages}
# source("https://raw.githubusercontent.com/miriamMota/scripts/master/installifnot.R")
source('E:/Dropbox (VHIR)/Scripts/usefulFunctions/installifnot.R')

installifnot("beeswarm")
installifnot("devtools")
# install_github(paste('alexsanchezpla/scripts/Rpackages/', "plotPCA2", sep="")) # 
# SI EN EL VHIR --> proxy=TRUE, ELSE proxy=FALSE
# installGitifnot("alexsanchezpla/scripts/Rpackages/","plotPCA2", force.install = TRUE, proxy=TRUE)
# installGitifnot("alexsanchezpla/scripts/Rpackages/","boxplotWithDots", force.install = TRUE, proxy=TRUE)
if(!(require(printr))) {
  install.packages(
    'printr',
    type = 'source',
    repos = c('http://yihui.name/xran', 'http://cran.rstudio.com')
  )
}
```

## Basic analyses

### Example inputs

For simplicity we have downloaded and stored in our `dades` directory the three input files mentioned above.

```{r}
fileName   <- "Datos.Normalizados.Filtrados.csv2"
HEAD <- TRUE
SEP=";"
DEC=","
ROWNAMES =1
myExpres <- read.table(file.path(dataDir, fileName), 
                     head=HEAD, sep=SEP, dec=DEC, row.names =ROWNAMES)
head(myExpres[,1:7])

targetsFileName <- "targets.txt"
SEP="\t"
DEC="."
targets <- read.table(file.path(dataDir, targetsFileName), head=TRUE, sep=SEP, dec=DEC)
head(targets)

topTableFileName <- "ExpressAndTop_AvsB.csv2"
SEP=";"
DEC=","
topTab <-  read.table(file.path(dataDir, topTableFileName), 
                      head=TRUE, sep=SEP, dec=DEC, row.names =1)
head(topTab[,1:6])
```

```{r eval=FALSE}
# This allows working with a "real" dataset
if (!exists("eset3noOutBalNoScan")) 
     load(file=file.path (dataDir, "results.ExpressionSets.Rda")) 
require(Biobase)

nabX <- t(exprs(eset3noOutBalNoScan))
dim(nabX)
nabY <- labs<-as.factor(pData(eset3noOutBalNoScan)$response)
table(nabY)
myExpres<-exprs(eset3noOutBalNoScan)
targets <- pData(eset3noOutBalNoScan)
lev<-as.factor(pData(eset3noOutBalNoScan)$response)
grupnames <- unique(lev)
```


### Exploring a single gene (1): Locating the gene

If one wants to work with the expression values -or the analysis values- of a given gene one must be able to locate it in the table containing these values. 

Imagine we want to explore gene "FOX1".

The topTable contains gene symbols so finding this gene is straightforward.
```{r}
geneSymbol <- "FOXA1"
geneIndex<- grep(geneSymbol, topTab$SymbolsA)
topTab[geneIndex,1:6]
```

The expression matrix does not have gene SYMBOLS as row names because there may be several probesets with the same symbol. In this case we need to get the probeset ID first, for instance from the tobTable (or if we don't have one, from an annotations table).
```{r}
geneSymbol <- "FOXA1"
geneIndex<- grep(geneSymbol, topTab$SymbolsA)
if (length(geneIndex) > 0)
  probesetName <- rownames(topTab)[geneIndex]
head(myExpres[probesetName,])
```

### Exploring a single gene (2): Drawing plots

Consider gene `FOXA1` the first one in the top table.

Start locating the gene as shown above
```{r}
geneName <- "FOXA1"
geneIndex<- grep(geneName, topTab$SymbolsA)
if (length(geneIndex) > 0)
  probesetName <- rownames(topTab)[geneIndex]
whichExpres<- myExpres[probesetName,]
whichExpres <- as.double(whichExpres)
whichGene <- which(rownames(myExpres)==probesetName)
whichInTopTab <- which(rownames(topTab)==probesetName)

# Alternatively one could do:
# whichGene <- which(rownames(myExpres)==probesetName)
# whichExpres
```

Another parameter that has to be set is the covariate that is used to define the groups.
```{r}
groupNames <- unique(targets$Group)
lev <- targets$Group
desc <- paste("logFC=", round(topTab$logFC[whichInTopTab],3), ", p-val=",
              round(topTab$P.Value[whichInTopTab],6),
              ", Adj-p=", round(topTab$adj.P.Val[whichInTopTab],6), sep="")
```

We may plot the expression of the selected gene for each group. This can be done using boxplots, dotplots or both plots overimposed as available in function \texttt{boxplotWithDots}.

```{r}
require(beeswarm)
geneName<-rownames(myExpres)[1]
whichExpres<-myExpres[1,]
opt<- par(mfrow=c(2,2)) # These calls should be turned into functions
beeswarm(whichExpres~lev, ylab="Expression", xlab="Groups", pch=13) 
boxplot(whichExpres~lev, ylab="Expression", xlab="Groups", pch=13,
         main=paste(geneName, desc, sep="\n"), labels=groupNames)  
beeswarm(whichExpres~lev, ylab="Expression", xlab="Groups", pch=13,
         main=paste(geneName, desc, sep="\n"), labels=groupNames) 
boxplot(whichExpres~lev, add = T, names = c("","",""), col="#0000ff22")
par(opt)
```

In this case it is better to only plot the points

```{r}
beeswarm(whichExpres~lev, ylab="Expression", xlab="Groups", pch=13,
         main=paste(geneName, desc, sep="\n"), labels=groupNames)  
```

### Looking at genes in pairs

We may want to see if the expression of the first three genes FOX1, VGLL1, SPDEF is related.

A simple approach is to use the `pairs` functions:

```{r}
idxs<- 1:3
geneNames <- as.character(topTab$SymbolsA[idxs])
probesetNames <- rownames(topTab)[idxs]
whichExpres <- myExpres[probesetNames,]
rownames(whichExpres)<- geneNames
pairs(t(whichExpres))
```

This may be improved to include the correlation coefficients in the upper panel:
```{r}
panel.cor <- function(x, y, digits = 2, cex.cor, ...)
{
  usr <- par("usr"); on.exit(par(usr))
  par(usr = c(0, 1, 0, 1))
  # correlation coefficient
  r <- cor(x, y)
  txt <- format(c(r, 0.123456789), digits = digits)[1]
  txt <- paste("r= ", txt, sep = "")
  text(0.5, 0.6, txt)

  # p-value calculation
  p <- cor.test(x, y)$p.value
  txt2 <- format(c(p, 0.123456789), digits = digits)[1]
  txt2 <- paste("p= ", txt2, sep = "")
  if(p<0.01) txt2 <- paste("p= ", "<0.01", sep = "")
  text(0.5, 0.4, txt2)
}
```

```{r}
idxs<- 1:3
geneNames <- as.character(topTab$SymbolsA[idxs])
probesetNames <- rownames(topTab)[idxs]
whichExpres <- myExpres[probesetNames,]
rownames(whichExpres)<- geneNames
pairs(t(whichExpres), upper.panel = panel.cor)
```


### A PCA plot

Principal components analysis is very useful to detect natural or technical groups (batches) in data.

```{r unPCA}
require(plotPCA2)
x <- myExpres
plotPCA2(X = x, colors = as.integer(targets$Group) + 1,  
         labels= targets$SampleIDs, transpose = TRUE)
legend('topright', c("A", "B", "L"), pch=1,  col= unique(as.integer(targets$Group)) + 1, bty='n')
```

## Analyzing _selected_ groups of genes 

Very often, post analyses are made on a list of genes which have been selected either by their biological relevance or because they verify some specific filtering condition.

For example imagine we wish to explore the most differentially expressed genes in the example topTable so we decide to select those genes that have an adjusted p-value of less than 0.001 and a minimum fold change of 2.

This selection can be done manually or using ad-hoc specific functions as described in the `geneLists` package [https://github.com/alexsanchezpla/scripts/tree/master/Rpackages/geneLists](https://github.com/alexsanchezpla/scripts/tree/master/Rpackages/geneLists).

```{r}
selectedTab<- topTab[abs(topTab$logFC)> 2 & topTab$adj.P.Val< 0.001,]
dim(selectedTab)
head(selectedTab)
```

The identifiers in the `selectedTab` object can now be used in different ways

- One can use the `Entrezs` identifiers to do some type of Gene Enrichment Analysis as described in the document 
- One can use the probeset names (the topTable rownames) to select the same rows in the expression matrix and perform some type of analysis such as a heatmap on the selected subset of genes.

Both possibilities are described in the final sections of the document [https://github.com/alexsanchezpla/scripts/blob/master/Exemple_Analisis_BioC/EjemploAnalisisMicroarrays.pdf](https://github.com/alexsanchezpla/scripts/blob/master/Exemple_Analisis_BioC/EjemploAnalisisMicroarrays.pdf).






